### PR TITLE
feat: grpc error handling for Map/MapT/MapStream/Reduce/Sink

### PR DESCRIPTION
**Pull request by xdevxy**
Closes [#769](https://github.com/numaproj/numaflow/issues/769)

This PR introduces a global GRPC interceptor for error handling (instead of handling cases one by one, which can be clumsy and easy to miss cases that needs to be handled). Although now exception can only happen at user custom code (application errors) which are translated to `UNKNOWN` grpc error code to be retried. The error handling logic can be extended easily to handle more type of exception if needed in future.

Also note that exception in reduce and sink now will no longer cause panic and exit but return with `UNKNOWN` grpc error to match the behavior in python and go sdk side.

### CODE REVIEW

File: src/main/java/io/numaproj/numaflow/function/FunctionServer.java [hunk 1]
```diff
@@ -151,7 +155,23 @@ public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
...
+                            handleException(ex, call, headers);
+                            throw ex;
+                        }
+                    }
+                    private void handleException(RuntimeException exception, ServerCall<ReqT, RespT> serverCall, Metadata headers) {
+                        // Currently, we only have application level exceptions.
+                        // Translate it to UNKNOWN status.
+                        serverCall.close(Status.UNKNOWN, headers);
+                    }
+                };
```
Comment (yhl25) [inline src/main/java/io/numaproj/numaflow/function/FunctionServer.java:174]: Why do we need this? Can't this be achieved using the onError() method of the observer?

File: src/main/java/io/numaproj/numaflow/function/FunctionServer.java [hunk 2]
```diff
@@ -151,7 +155,23 @@ public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
...
+                        } catch (RuntimeException ex) {
+                            handleException(ex, call, headers);
+                            throw ex;
+                        }
+                    }
+                    private void handleException(RuntimeException exception, ServerCall<ReqT, RespT> serverCall, Metadata headers) {
+                        // Currently, we only have application level exceptions.
+                        // Translate it to UNKNOWN status.
+                        serverCall.close(Status.UNKNOWN, headers);
+                    }
```
Comment (yhl25) [inline src/main/java/io/numaproj/numaflow/function/FunctionServer.java:173]: what headers are we sending back?

File: src/main/java/io/numaproj/numaflow/function/FunctionService.java [hunk 3]
```diff
@@ -296,14 +296,13 @@ private Udfunction.DatumResponseList buildDatumListResponse(MessageTList message
...
     private void handleFailure(CompletableFuture<Void> failureFuture) {
         new Thread(() -> {
             try {
                 failureFuture.get();
             } catch (Exception e) {
-                e.printStackTrace();
-                System.exit(1);
+                throw new RuntimeException("error in reduce fn", e);
             }
         }).start();
```
Comment (yhl25) [inline src/main/java/io/numaproj/numaflow/function/FunctionService.java:307]: Why are we throwing the exception inside the thread? it has no effect

File: src/main/java/io/numaproj/numaflow/sink/SinkService.java [hunk 4]
```diff
@@ -88,16 +88,13 @@ public void isReady(Empty request, StreamObserver<Udsink.ReadyResponse> response
...
+    // wrap the exception and let it be handled in the central error handling logic.
     private void handleFailure(CompletableFuture<Void> failureFuture) {
         new Thread(() -> {
             try {
                 failureFuture.get();
             } catch (Exception e) {
-                e.printStackTrace();
-                System.exit(1);
+                throw new RuntimeException("error in sinker fn", e);
             }
```
Comment (yhl25) [inline src/main/java/io/numaproj/numaflow/sink/SinkService.java:98]: same comment

File: src/test/java/io/numaproj/numaflow/function/FunctionServerTestErr.java [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+                        .addKeys(reduceKey)
+                        .build();
+                inputStreamObserver.onNext(inputDatum);
+            }
+
+            inputStreamObserver.onCompleted();
+        } catch (Exception e) {
+            assertEquals(Status.UNKNOWN.getCode().toString(), e.getMessage());
+        }
+    }
```
Comment (yhl25) [inline src/test/java/io/numaproj/numaflow/function/FunctionServerTestErr.java:180]: is this part even getting executed?

File: src/test/java/io/numaproj/numaflow/sink/SinkServerTestErr.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                        .addAllKeys(List.of(keys))
+                        .build();
+                inputStreamObserver.onNext(inputDatum);
+            }
+
+            inputStreamObserver.onCompleted();
+        } catch (Exception e) {
+            assertEquals(Status.UNKNOWN.getCode().toString(), e.getMessage());
+        }
+    }
```
Comment (yhl25) [inline src/test/java/io/numaproj/numaflow/sink/SinkServerTestErr.java:89]: same comment

File: src/main/java/io/numaproj/numaflow/function/FunctionServer.java (hunk 1 above)
Comment (xdevxy) [reply src/main/java/io/numaproj/numaflow/function/FunctionServer.java:174]: Using the onError() works but it requires specific error handling in each case there is exception. This is to use a GRPC interceptor to build a central place for error handling, so any new exception can be added and handled here without adding extra logic in other places.

File: src/main/java/io/numaproj/numaflow/function/FunctionServer.java (hunk 2 above)
Comment (xdevxy) [reply src/main/java/io/numaproj/numaflow/function/FunctionServer.java:173]: In the interceptor we are not adding anything to the headers. the original headers are sending back.

Comment (yhl25) [reply src/main/java/io/numaproj/numaflow/function/FunctionServer.java:173]: Why do we need to send them back?

File: src/main/java/io/numaproj/numaflow/function/FunctionService.java (hunk 3 above)
Comment (xdevxy) [reply src/main/java/io/numaproj/numaflow/function/FunctionService.java:307]: As documented in L299. Here we wrap all the exceptions happens at the custom user code (application layer) as a `RuntimeException` and let it handled in the global grpc interceptor.

File: src/main/java/io/numaproj/numaflow/sink/SinkService.java (hunk 4 above)
Comment (xdevxy) [reply src/main/java/io/numaproj/numaflow/sink/SinkService.java:98]: Same above.

File: src/main/java/io/numaproj/numaflow/function/FunctionService.java (hunk 3 above)
Comment (yhl25) [reply src/main/java/io/numaproj/numaflow/function/FunctionService.java:307]: since we are throwing the exception inside the thread, the global grpc interceptor won't get the exception.

File: src/test/java/io/numaproj/numaflow/function/FunctionServerTestErr.java (hunk 5 above)
Comment (xdevxy) [reply src/test/java/io/numaproj/numaflow/function/FunctionServerTestErr.java:180]: yes, it assert if the status code is `unknown`.

Comment (yhl25) [reply src/test/java/io/numaproj/numaflow/function/FunctionServerTestErr.java:180]: is the catch block getting executed?

Comment (xdevxy) [reply src/test/java/io/numaproj/numaflow/function/FunctionServerTestErr.java:180]: yes

File: src/main/java/io/numaproj/numaflow/function/FunctionService.java (hunk 3 above)
Comment (xdevxy) [reply src/main/java/io/numaproj/numaflow/function/FunctionService.java:307]: Hmm, but I wrote the test to ensure the grpc interceptor indeed get the exception.

File: src/main/java/io/numaproj/numaflow/function/FunctionServer.java (hunk 2 above)
Comment (xdevxy) [reply src/main/java/io/numaproj/numaflow/function/FunctionServer.java:173]: LOL, this is part of required parameters for `serverCall.close()`

File: src/main/java/io/numaproj/numaflow/function/FunctionServer.java (hunk 1 above)
Comment (xdevxy) [reply src/main/java/io/numaproj/numaflow/function/FunctionServer.java:174]: I think your questions is orthogonal to whether using onError directly or interceptor. More for how to handle the errors then. For map, the error handling case isn't for application error inside the UDF? And for reduce, we need to send back error for the client to say if retry is needed to match go/python sdk behavior.

File: src/main/java/io/numaproj/numaflow/function/FunctionServer.java (hunk 2 above)
Comment (yhl25) [reply src/main/java/io/numaproj/numaflow/function/FunctionServer.java:173]: Sorry, let me rephrase my question. Why are we sending the input headers back? Is it used on the client side? 
I know it's a required parameter, we can just go with null right?

Comment (xdevxy) [reply src/main/java/io/numaproj/numaflow/function/FunctionServer.java:173]: Got it, I don't see any reason we would like to update the original headers or send a null instead as you can see we are even adding datum metadata information to the headers at L155-157. Do you have any particular reason in mind?

File: src/main/java/io/numaproj/numaflow/function/FunctionService.java (hunk 3 above)
Comment (xdevxy) [reply src/main/java/io/numaproj/numaflow/function/FunctionService.java:307]: Sorry, now I re-read your previous comment, I got your point. In this case, it seems we should handle this specifically in the `ReduceShutdownActor.shutdown()`? Also, I don't see a good way to test this?

File: src/main/java/io/numaproj/numaflow/function/FunctionServer.java (hunk 2 above)
Comment (yhl25) [reply src/main/java/io/numaproj/numaflow/function/FunctionServer.java:173]: No reason, just wanted to know how it was used.

File: src/main/java/io/numaproj/numaflow/sink/SinkServer.java [hunk 7]
```diff
@@ -95,8 +101,33 @@ public void start() throws Exception {
...
+                        }
+                    }
+                    private void handleException(RuntimeException exception, ServerCall<ReqT, RespT> serverCall, Metadata headers) {
+                        // Currently, we only have application level exceptions.
+                        // Translate it to UNKNOWN status.
+                        serverCall.close(Status.UNKNOWN, headers);
+                    }
+                };
+            }
+        };
```
Comment (yhl25) [inline src/main/java/io/numaproj/numaflow/sink/SinkServer.java:127]: Do we still need this for sink?

File: src/test/java/io/numaproj/numaflow/function/FunctionServerTestErr.java [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+
+        try {
+            new Thread(() -> {
+                while (outputStreamObserver.t == null);
+                assertEquals("UNKNOWN: java.lang.RuntimeException: unknown exception", outputStreamObserver.t.getMessage());
+            }).join();
+        } catch (InterruptedException e) {
+            fail("Thread interrupted");
+        }
+
```
Comment (yhl25) [inline src/test/java/io/numaproj/numaflow/function/FunctionServerTestErr.java:175]: how are making sure that this thread gets executed?

File: src/test/java/io/numaproj/numaflow/function/FunctionServerTestErr.java [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+                .reduceFn(outputStreamObserver);
+        for (int i = 1; i <= 10; i++) {
+            Udfunction.DatumRequest inputDatum = Udfunction.DatumRequest.newBuilder()
+                    .setValue(ByteString.copyFromUtf8(String.valueOf(i)))
+                    .addKeys(reduceKey)
+                    .build();
+            inputStreamObserver.onNext(inputDatum);
+        }
+
+        inputStreamObserver.onCompleted();
```
Comment (yhl25) [inline src/test/java/io/numaproj/numaflow/function/FunctionServerTestErr.java:188]:
```suggestion
        StreamObserver<Udfunction.DatumRequest> inputStreamObserver = UserDefinedFunctionGrpc
                .newStub(inProcessChannel)
                .withInterceptors(MetadataUtils.newAttachHeadersInterceptor(metadata))
                .reduceFn(outputStreamObserver);
        for (int i = 1; i <= 10; i++) {
            Udfunction.DatumRequest inputDatum = Udfunction.DatumRequest.newBuilder()
                    .setValue(ByteString.copyFromUtf8(String.valueOf(i)))
                    .addKeys(reduceKey)
                    .build();
            inputStreamObserver.onNext(inputDatum);
        }

        inputStreamObserver.onCompleted();
        while (outputStreamObserver.t == null);
        assertEquals("UNKNOWN: java.lang.RuntimeException: unknown exception", outputStreamObserver.t.getMessage());
```

**Review by yhl25 — APPROVED** [review]
LGTM
